### PR TITLE
feat(developer): use kmc to compile project files 🎺

### LINKED ISSUE #7688: bug(developer): compiler fails to build files in non-ascii paths
* **TEST_NON_ASCII_PATHS (FAILED):** Retested this PR with Keyman 16.0.98-beta-test-7631 build in Windows 10 OS and here is my observation. 1. I was able to create project (in Khmer and Tamil) within that folder which contains Non-Ascii letters. 2. I did not see any error message in Dev Console (after pressing F12). But, I was getting 'Fatal Error' message after compiling a project. Seems to be an issue. 

..Khmer Project

   

..Tamil Project

   

_Originally posted by @bharanidharanj in https://github.com/keymanapp/keyman/issues/7631#issuecomment-1309896697_

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #7688.

Uses kmc to compile .kmn, .model.ts, and .kps files in TIKE.

Also:
* Consolidates the outFile parameter that was duplicated in kmc-kmn.
* removes unused FSilent flag (was always False).

# User Testing

## SUITE_KEYBOARD: Compiling keyboards

* **TEST_COMPILE_KEYBOARD**: Open several existing keyboard projects in Keyman Developer and try and compile each keyboard, from the Project window. Verify that the outputs correspond to 16.0 release outputs.

* **TEST_COMPILE_KEYBOARD_EDITOR**: Open several existing keyboard projects in Keyman Developer and try and compile each keyboard, from the Keyboard Editor Compile tab. Verify that the outputs correspond to 16.0 release outputs.

## SUITE_PACKAGE: Compiling packages

* **TEST_COMPILE_PACKAGE**: Open several existing keyboard projects in Keyman Developer and try and compile each package, from the Project window (you may need to compile the keyboards first). Verify that the outputs correspond to 16.0 release outputs. Try also with model projects.

* **TEST_COMPILE_PACKAGE_EDITOR**: Open several existing keyboard projects in Keyman Developer and try and compile each package, from the Package Editor Compile tab (you may need to compile the keyboards first). Verify that the outputs correspond to 16.0 release outputs. Try also with model projects.

## SUITE_VARIOUS: Various other tasks

* **TEST_NON_ASCII_PATHS**: Create a folder that contains non-ASCII letters, e.g. Khmer, Tamil or European diacritics. Create a new keyboard project within that folder. In the Project view, press F12 to verify that no errors are raised in the Developer console. Do a basic regression test on working with the keyboard project, including editing, compiling (from #7688).

### CODE REVIEW

File: developer/src/tike/project/Keyman.Developer.System.Project.kmnProjectFileAction.pas [hunk 1]
```diff
@@ -80,8 +46,8 @@ function TkmnProjectFileAction.Clean: Boolean;
   begin
     FJS := TKeyboardUtils.GetKeymanWebCompiledFileName(FileName);
     CleanFile(FJS); // keyboard-x.y.js
-    CleanFile(ChangeFileExt(FJS, '') + '_load.js'); // keyboard-x.y_load.js
-    CleanFile(ChangeFileExt(FJS, '.json'), True); // keyboard-x.y_load.js
+//    CleanFile(ChangeFileExt(FJS, '') + '_load.js'); // keyboard-x.y_load.js
+//    CleanFile(ChangeFileExt(FJS, '.json'), True); // keyboard-x.y_load.js
```
Comment (ermshiperete) [inline developer/src/tike/project/Keyman.Developer.System.Project.kmnProjectFileAction.pas:50]: So we want to keep these files now?

**Review by ermshiperete — APPROVED** [review]

File: developer/src/tike/project/Keyman.Developer.System.Project.kmnProjectFileAction.pas (hunk 1 above)
Comment (mcdurdin) [reply developer/src/tike/project/Keyman.Developer.System.Project.kmnProjectFileAction.pas:50]: Those files are no longer generated -- they are both obsolete. I should probably remove the code altogether.

```suggestion
```

File: developer/src/tike/compile/Keyman.Developer.System.KmcWrapper.pas [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  msgText: string;
+  msgCode: Integer;
+  msgLine: Integer;
+  msgType: string;
+begin
+  ec := 0;
+  logtext := '';
+
+  // TODO: log-level as opt?
+  cmdline := Format('"%s" build --log-format tsv --log-level info "%s"', [TKeymanDeveloperPaths.KmcPath, infile]);
```
[the remark this answers is not in the document]
Comment (mcdurdin) [reply developer/src/tike/compile/Keyman.Developer.System.KmcWrapper.pas:50]: The `debug` parameter here is for the compiler to generate debug version of the compiled keyboard; it's not about debugging the compiler itself :grin: so it's a little confusing I admit.

File: developer/src/kmc/src/commands/buildClasses/BuildKmnKeyboard.ts [hunk 3]
```diff
@@ -16,14 +17,18 @@ export class BuildKmnKeyboard extends BuildActivity {
...
     // We need to resolve paths to absolute paths before calling kmc-kmn
-    let outfile = this.getOutputFilename(infile, options);
+    if(options.outFile) {
+      options.outFile = getPosixAbsolutePath(options.outFile);
+    }
 
     infile = getPosixAbsolutePath(infile);
-    outfile = getPosixAbsolutePath(outfile);
-
-    // TODO: Currently this only builds .kmn->.kmx, and targeting .js is as-yet unsupported
```
[the remark this answers is not in the document]
Comment (mcdurdin) [reply developer/src/kmc/src/commands/buildClasses/BuildKmnKeyboard.ts:24]: Yes, the compiler builds both .kmx and .js from a .kmn, as determined by the `&targets` store.

**Comment by bharanidharanj** [comment]
## SUITE_KEYBOARD: Compiling keyboards

* **TEST_COMPILE_KEYBOARD (PASSED):** Here, what I did is: 1. Installed Keyman Developer 16.0 Stable build on Windows 11 OS (VM). 2. Opened several existing projects then compiled them. 3. Uninstalled 16.0 and Installed the attached Keyman Developer 17.0.159-alpha-test-7539 build on Windows 11 OS. 4. Opened the existing projects then compiled them. 5. Verified the outputs corresponding to 16.0 release outputs and it seems that the two results are distinct and exceptional.

..sil_euro_latin_project

   


..khmerangkor project

   


..itransdevanagiri project

   


..gff_amharic project

**Comment by bharanidharanj** [comment]
# Test Results

## SUITE_PACKAGE: Compiling packages

* **TEST_COMPILE_PACKAGE (PASSED):** Here, what I did is: 1. Installed Keyman Developer 16.0 Stable build on Windows 11 OS (VM). 2. Opened several existing projects and then compiled them via the keyboard project window. 3. Uninstalled 16.0 and installed the attached Keyman Developer 17.0.159-alpha-test-7539 build on Windows 11 OS. 4. Opened the existing projects and then compiled them. 5. Verified the outputs corresponding to 16.0 release outputs; the two results seem distinct and exceptional.


..gff_amharic

   


..sil_eurolatin

   

..itransdevanagiri

**Comment by bharanidharanj** [comment]
## SUITE_PACKAGE: Compiling packages

* **TEST_COMPILE_PACKAGE_EDITOR (PASSED):** Here, what I did is: 1. Installed Keyman Developer 16.0 Stable build on Windows 11 OS (VM). 2. Opened several existing projects then compiled them via package editor window. 3. Uninstalled 16.0 and installed the attached Keyman Developer 17.0.159-alpha-test-7539 build on Windows 11 OS. 4. Opened the existing projects and then compiled them. 5. Verified the outputs corresponding to 16.0 release outputs; the two results seem distinct and exceptional.

..gff_amharic

   


..sil_eurolatin

   


..itransdevanagiri

**Comment by bharanidharanj** [comment]
# Test Results

## SUITE_VARIOUS: Various other tasks

* **TEST_NON_ASCII_PATHS (PASSED):** Tested with the attached PR build (Keyman Developer 17.0.159-alpha-test-9539) in Windows 11 (VM) and here is my observation: 1. Created a new folder that contains non-ASCII letters (I have tried with Tamil and Khmer language). 2. Created a new keyboard project within that folder. In the project view, I pressed F12 to see the Dev Console without showing any errors. 3. Edited the Keyboard project by including some letters in the keyboard layout view. Also, inserted some long-press keys in the Touch layout view. Then, compiled the project. I did not see any error message after the compilation. Seems to be working as expected.

**Pull request merged**